src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs

```diff
@@ -0,0 +1,377 @@
+                        // kind as the current edit.
+                        editKind = currentEditKind;
+                    }
+
+                    if (isCurrentEditKindUpdate || currentEditKind == SemanticTokenEditKind.Insert)
+                    {
+                        tokensToInsert.AddRange(newGroupedSemanticTokens[editIndexes[i + 1]].ToArray());
+                    }
+
+                    if (isCurrentEditKindUpdate || currentEditKind == SemanticTokenEditKind.Delete)
+                    {
+                        deleteCount += 5;
+                    }
+
+                    i++;
+                }
+
+                semanticTokensEdits.Add(
+                    GenerateEdit(start: initialEditIndex * 5, deleteCount: deleteCount, data: tokensToInsert.ToArray()));
+
+                return i;
+            }
+
+            static int AddInsertionEdit(
+                SemanticToken[] newGroupedSemanticTokens,
+                Dictionary<int, SemanticTokenEditKind> indexToEditKinds,
+                ArrayBuilder<SemanticTokensEdit> semanticTokensEdits,
+                int[] editIndexes,
+                int i,
+                int initialEditIndex)
+            {
+                var _ = ArrayBuilder<int>.GetInstance(out var tokensToInsert);
+                tokensToInsert.AddRange(newGroupedSemanticTokens[initialEditIndex].ToArray());
+
+                // An insert can only be combined with other inserts that directly follow it.
+                while (i + 1 < editIndexes.Length && indexToEditKinds[editIndexes[i + 1]] == SemanticTokenEditKind.Insert &&
+                    editIndexes[i + 1] == editIndexes[i] + 1)
+                {
+                    tokensToInsert.AddRange(newGroupedSemanticTokens[editIndexes[i + 1]].ToArray());
+                    i++;
+                }
+
+                semanticTokensEdits.Add(
+                    GenerateEdit(start: initialEditIndex * 5, deleteCount: 0, data: tokensToInsert.ToArray()));
+                return i;
+            }
+
+            static int AddDeletionEdit(
+                Dictionary<int, SemanticTokenEditKind> indexToEditKinds,
+                ArrayBuilder<SemanticTokensEdit> semanticTokensEdits,
+                int[] editIndexes,
+                int i,
+                int initialEditIndex)
+            {
+                var deleteCount = 5;
+
+                // A deletion can only be combined with other deletions that directly follow it.
+                while (i + 1 < editIndexes.Length && indexToEditKinds[editIndexes[i + 1]] == SemanticTokenEditKind.Delete &&
+                    editIndexes[i + 1] == editIndexes[i] + 1)
+                {
+                    deleteCount += 5;
+                    i++;
+                }
+
+                semanticTokensEdits.Add(
+                    GenerateEdit(start: initialEditIndex * 5, deleteCount: deleteCount, data: Array.Empty<int>()));
+                return i;
+            }
```

Comment:
my original thought was that you could add these three methods together.  however, it's prob nice to keep them separate and nice and clean.

That said, the methods for insert/delete are really simple and easy to understand.  have you considered just tighting the rule to "you can only merge a sequence of all-updates, a sequence of all-deletes, or a seqeunce of all-inserts"?

That would simplify the first method greatly.  yes, we wouldnt' merge a couple of cases (like update-update-update-insert-insert-insert).   However, those would still get broken into sequences that would get merged.  So that seems fine.

Reply:
That's a good point, the logic would be definitely simpler if we tightened the rules. I have updated the code to reflect your suggestion. Thank you!